src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -251,20 +252,197 @@ export class MagicWandService {
   }
 
   /**@license MIT License <http://www.opensource.org/licenses/mit-license.php>
-  * Pretty good color distance from
-  * http://www.compuphase.com/cmetric.htm
-  */
+   * Pretty good color distance from
+   * http://www.compuphase.com/cmetric.htm
+   */
   colorDistance(a: Color, b: Color): number {
     const dr = a.red - b.red;
     const dg = a.green - b.green;
     const db = a.blue - b.blue;
     const redMean = (a.red + b.red)/2;
     return (2+redMean/256)*dr*dr + 4*dg*dg + (2 + (255 - redMean)/256)*db*db;
   }
+
+
+  /* Code for 'preview' algorithm */
+
+  comparator(pixelA: PixelNode, pixelB: PixelNode): boolean {
+    return pixelA.distance < pixelB.distance;
+  }
+
+  /**Implements Dijkstra's algorithm to compute the shortest distance
+   * from the original, inputted pixel to every other pixel.
+   * The distance along a path's colorDistance is determined as the 
+   * maximum colorDistance(in relation to the input pixel) of all pixels
+   * in that path.
+   * @returns {Array<number>} an array of the computed distances.
+   */
+  getShortestPaths(imgData: ImageData, xCoord: number, yCoord: number,
+      toleranceLimit: number): Array<number> {
+    // Stores a queue of coords for pixels that we need to visit in "visit".
+    const toVisit: PriorityQueue<PixelNode> = 
+        new PriorityQueue(this.comparator);
+    const distances: Array<number> = [];
+    distances.fill(undefined, 0, imgData.data.length / 4 + 1);
+
+    // Stores already-visited pixels in "visited" as index formatted numbers
+    // (as opposed to coord format; for Set funcs).
+    const visited: Set<number> = new Set();
+
+    const originalPixelColor: Color = 
+        this.dataArrayToRgb(imgData, xCoord, yCoord);
+
+    const originalIndex: number = 
+        this.coordToDataArrayIndex(xCoord, yCoord, imgData.width);
+    toVisit.push({
+        distance: 0,
+        index: originalIndex / 4});
+
+    distances[originalIndex / 4] = 0;
+
+    // Works with tolerance limit in the squared space.
+    toleranceLimit *= toleranceLimit;
+
+    // Updates shortest path between neighbor pixel and original node.
+    while (toVisit.getSize() !== 0) {
+      const curPixelNode: PixelNode = toVisit.pop();
+      // A node is considered visited once popped.
```

Comment:
I actually think you want to consider a node visited once you push it into toVisit. Otherwise you could end up visiting the same node multiple times, which is redundant. (With this particular distance metric, unlike the general case, you will always have found the shortest path the first time you encounter a neighbor.)

Reply:
Ahhh I see, thanks! That should help a lot with keeping the size of the priority queue, and thus the time complexity of push and pop, optimized.